src/DynamoCoreWpf/Rendering/HelixRenderPackage.cs

```diff
@@ -555,12 +521,281 @@ public IEnumerable<int> PointIndices
+                }
+                
+                return list;
+            }
+        }
+
+        /// <summary>
+        /// Add a label position to the render package.
+        /// </summary>
+        public void AddLabel(string label, VertexType vertexType, int index)
+        {
+            Vector3 position;
+            switch (vertexType)
+            {
+                case VertexType.Point:
+                    if (index > points.Positions.Count) {return;}
+                    position = points.Positions[index-1];
+                    break;
+                case VertexType.Line:
+                    if (index > lines.Positions.Count) { return; }
+                    position = lines.Positions[index - 1];
+                    break;
+                case VertexType.Mesh:
+                    if (index > mesh.Positions.Count) { return; }
+                    position = Mesh.Positions[index - 1];
+                    break;
+                default:
+                    return;
+            }
+            LabelPlaces.Add(new Tuple<string, Vector3>(label,position));
+        }
+
+        /// <summary>
+        /// Add a label position to the render package.
+        /// </summary>
+        public void AddLabel(string label, double x, double y, double z)
+        {
+            LabelPlaces.Add(new Tuple<string, Vector3>(label, Vector3ForYUp(x, y, z)));
+        }
+
+        /// <summary>
+        /// A flag indicating whether the render package should auto generate labels
+        /// </summary>
+        public bool AutoGenerateLabels { get; set; } = true;
+
+        /// <summary>
+        /// Clear all label data from the render package.
+        /// </summary>
+        public void ClearLabels()
+        {
+            LabelPlaces.Clear();
+            AutoGenerateLabels = true;
+        }
+
+        #endregion
+
+        #region IRenderPackageSupplement implementation
+
+        /// <summary>
+        /// The number of point vertices colors in the package.
+        /// </summary>
+        public int PointVertexColorCount => points.Colors.Count;
+
+        /// <summary>
+        /// The number of line vertices colors in the package.
+        /// </summary>
+        public int LineVertexColorCount => lines.Colors.Count;
+
+        /// <summary>
+        /// The number of mesh vertices colors in the package.
+        /// </summary>
+        public int MeshVertexColorCount => mesh.Colors.Count;
+
+        /// <summary>
+        /// Insert a color to a range of point vertices.
+        /// </summary>
+        public void InsertPointVertexColorRange(int startIndex, int endIndex, byte red, byte green, byte blue, byte alpha)
+        {
+            var message = string.Empty;
+            if (!HasValidStartEnd(startIndex, endIndex, points, out message))
+            {
+                if (!string.IsNullOrEmpty(message))
+                {
+                    throw new Exception(message);
+                }
```

Comment:
Why throw an exception only if there is a message? In any case I think the message existing and `HasValidStartEnd` returning false go hand-in-hand.